### PR TITLE
Implement encryption of file

### PR DESCRIPTION
**Pull request by oleksiivorobiov**

### CODE REVIEW

File: src/mongoc/mongoc-gridfs-cnv-file.c [hunk 1]
```diff
@@ -167,9 +222,29 @@ mongoc_gridfs_cnv_file_destroy (mongoc_gridfs_cnv_file_t *file)
...
+   bson_iter_binary (&it, NULL, &saved_after_encryption_tag_len, &saved_after_encryption_tag);
+   BSON_ASSERT (saved_after_encryption_tag_len == AES_BLOCK_SIZE);
+
+   BSON_ASSERT (eax_compute_tag(tag, sizeof tag, &file->aes_ctx) == RETURN_GOOD);
+
+   return memcmp (tag, saved_after_encryption_tag, AES_BLOCK_SIZE) == 0;
+}
+
 ssize_t
 mongoc_gridfs_cnv_file_readv (mongoc_gridfs_cnv_file_t *file,
```
Comment (jsbattig) [inline src/mongoc/mongoc-gridfs-cnv-file.c:249]: @alexeyvo , the reason you had to write a custom file_readv is mainly to check the status of flags calculated during the decompression/decrypt of chunk?

File: src/mongoc/mongoc-gridfs-cnv-file.c [hunk 2]
```diff
@@ -254,3 +361,27 @@ mongoc_gridfs_cnv_file_get_compressed_length (mongoc_gridfs_cnv_file_t *file)
 {
    return file->compressed_length;
 }
+
+bool
+mongoc_gridfs_cnv_file_set_aes_key (mongoc_gridfs_cnv_file_t *file,
+                                    const unsigned char      *aes_key,
```
Comment (jsbattig) [inline src/mongoc/mongoc-gridfs-cnv-file.c:367]: This is an actual AES key right? not a passphrase?

Comment (jsbattig) [inline src/mongoc/mongoc-gridfs-cnv-file.c:367]: I can answer my own question: yes… I see the code below that checks the key size to make sure it's a valid AES key.

File: src/mongoc/mongoc-gridfs-cnv-file.c (hunk 1 above)
Comment (oleksiivorobiov) [inline src/mongoc/mongoc-gridfs-cnv-file.c:249]: I use object composition, so I call mongoc_gridfs_file_readv inside mongoc_gridfs_cnv_file_readv, I do the same for all functions even if there is no additional logic. file->flags are set on object creation and I just check If we try to decrypt without set aes key - it's reading error. If read() returns 0 bytes it means we read all data and we should check data integrity using EAX tag.

File: src/mongoc/mongoc-gridfs-cnv-file.c (hunk 2 above)
Comment (oleksiivorobiov) [inline src/mongoc/mongoc-gridfs-cnv-file.c:367]: Yes I think we can have ability to set key directly in addition to use passphrase.

**Pull request merged**